### DISCUSSION TITLE
can't explain the extrusion direction of a beam

### DISCUSSION THREAD

**Discussion by zooid** [IfcOpenShell/IfcOpenShell#4937] [Q&A] [answered]
I've been struggling with this for a few days now I just can't figure out why my 2nd ifcbeam object is extruded in the wrong direction. I'm sure there's a simple explanation but with my limited knowledge of IFC in general and ifcopenshell I can't figure it out alone.

Could anyone help out?

here's link to the python file gist
[<script src="https://gist.github.com/zooid/4ff4b7c2900765137fee222bcae1a668.js"></script>](https://gist.github.com/zooid/4ff4b7c2900765137fee222bcae1a668)

This is how the ifc file looks. 
![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/1837230/6911a003-f12a-4072-9ed4-2ee1be7ab387)

**Comment by aothms** [comment, marked as answer]
```
    profile_plane = _model.create_entity("IfcAxis2Placement3D",
                                    Location=position_point,
                                    Axis=extrusion_direction,
                                    RefDirection=local_x_axis)

    extruded_area_solid = _model.create_entity("IfcExtrudedAreaSolid",
                                               SweptArea=_profile,
                                               Position=profile_plane,
                                               ExtrudedDirection=extrusion_direction,
                                               Depth=_extrusion_length)
```

^ I just scrolled through the code, so didn't test but you're 'applying' extrusion_direction twice. You should think of the Position as a post-hoc transformation after extruding, not just as defining in plane independent of the the extrusion vector. The extrusion vector instead 'moves along' with the placement in Position.

So, all things considered I would try setting `ExtrudedDirection=_model.create_entity("IfcDirection", (0., 0., 1.))`.

Even better, what you also started attempting is 

```
Position=None,
ExtrudedDirection=_model.create_entity("IfcDirection", (0., 0., 1.)),
```

and uncomment `RelativePlacement=profile_plane`.

**Comment by Moult** [comment]
In IFC, the extrusion direction of a beam, column, or any extrusion should be 0, 0, 1. To make it point any other direction, instead you should rotate the object placement of the beam.

Might I also recommend using the API? It will be much, much simpler and avoid other mistakes :) In particular see: https://docs.ifcopenshell.org/ifcopenshell-python/code_examples.html#create-a-simple-model-from-scratch

Then see https://docs.ifcopenshell.org/ifcopenshell-python/geometry_creation.html in particular profile extrusions https://docs.ifcopenshell.org/ifcopenshell-python/geometry_creation.html#profile-representations

**Reply by zooid** [reply]
Thanks @aothms this worked! :)
I hope it will work on my complete program :).

**Reply by zooid** [reply]
@Moult, the examples given in the documentation do not work, I had to use ChatGPT during a few days in order to make sense of both the documentation and the validation errors at https://validate.buildingsmart.org/.

I think complete examples from scratch would be very useful, but also explaining the topology of IFC because, right now you run an example and it either won't create the IFC file, or it will create it but you can view it in BIMVision but you can't in Revit or you can't in both.

**Reply by Moult** [reply]
What errors are you facing with the examples?

**Reply by zooid** [reply]
at first, there were examples where there was function `run` without the import directive so I could not understand where it is coming from. 
Then it was the `ifcopenshell.api` neither PyCharm nor Jupyter could find the functions from the examples.
I asked ChatGPT with some explanation and the code from documentation and it spit out something that kinda worked but didn't create IFC, then after a few iterations it did create the file but nothing was visible in BIMVision. I could not find information what should IfcBeam be put in. ChatGPT helped with that too. 
Finally, my beams were visible in BIMvision but not in Revit, so I had to search for hours to figure out what does Revit require, and it was yet another object in the IFC topology, I don't remember which was it and I deleted my iterations they were hundreds.

So, all in all, it's a combination, both not enough information about the functions and how they are related + the IFC required topology to have something work with Revit. 

I'm usually working with 3DEXPERIENCE and it has a terrible IFC exporter, that neither Revit nor Tekla can read. Now I understand why.

**Reply by zooid** [reply]
example validation error:
```
<html>
<body>
<!--StartFragment--><div class="css-tc4u2w MuiTreeItem-content Mui-expanded Mui-selected" style="padding: 0px 8px; width: 850px; box-sizing: border-box; display: flex; -webkit-box-align: center; align-items: center; cursor: pointer; -webkit-tap-highlight-color: transparent; background-color: rgba(25, 118, 210, 0.08); border-bottom: 1px solid black; color: rgba(0, 0, 0, 0.87); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Roboto, Oxygen, Ubuntu, Cantarell, &quot;Fira Sans&quot;, &quot;Droid Sans&quot;, &quot;Helvetica Neue&quot;, sans-serif; font-size: medium; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: left; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><div class="MuiTreeItem-label" style="padding-left: 4px; width: 815px; box-sizing: border-box; min-width: 0px; position: relative; font-family: Roboto, Helvetica, Arial, sans-serif; font-weight: 400; font-size: 1rem; line-height: 1.5; letter-spacing: 0.00938em;"><div><div class="caption" style="padding-top: 1em; padding-bottom: 1em; text-transform: capitalize;">Entity Rule<span> </span>-<span> </span>IfcProduct.PlacementForShapeRepresentation</div><div class="subcaption" style="visibility: visible; font-size: 12.8px;">(exists(representation) and exists(objectplacement) or (exists(representation) and sizeof([temp for temp in representation.Representations if 'ifc4.ifcshaperepresentation' in typeof(temp)]) == 0) or (not exists(representation)))</div></div></div></div><ul class="MuiCollapse-root MuiCollapse-vertical MuiTreeItem-group MuiCollapse-entered css-1xxsnna" role="group" style="height: auto; overflow: visible; transition: height 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; margin: 0px; padding: 16px; color: rgba(0, 0, 0, 0.87); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Roboto, Oxygen, Ubuntu, Cantarell, &quot;Fira Sans&quot;, &quot;Droid Sans&quot;, &quot;Helvetica Neue&quot;, sans-serif; font-size: medium; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: left; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial; min-height: 0px;"><div class="MuiCollapse-wrapper MuiCollapse-vertical css-hboir5" style="display: flex; width: 818px;"><div class="MuiCollapse-wrapperInner MuiCollapse-vertical css-8atqhb" style="width: 818px;">
Id | Entity | Severity | Message
-- | -- | -- | --
#29 | IfcBeam | Error | Violated by:     (((True and False) or (True and 1 == 0) or not True))      +  where True = exists(#28=IfcProductDefinitionShape($,$,(#27)))      +  and   False = exists(None)      +  and   True = exists(#28=IfcProductDefinitionShape($,$,(#27)))      +  and   1 = sizeof([#27=IfcShapeRepresentation(#9,'Body','SweptSolid',(#26))])      +  and   True = exists(#28=IfcProductDefinitionShape($,$,(#27)))  On instance:     #29=IfcBeam('2PptWzwh5BVwHzB2opxFJu',$,'Example Beam Rectangle',$,'Beam',$,#28,$,$)
#40 | IfcBeam | Error | Violated by:     (((True and False) or (True and 1 == 0) or not True))      +  where True = exists(#39=IfcProductDefinitionShape($,$,(#38)))      +  and   False = exists(None)      +  and   True = exists(#39=IfcProductDefinitionShape($,$,(#38)))      +  and   1 = sizeof([#38=IfcShapeRepresentation(#9,'Body','SweptSolid',(#37))])      +  and   True = exists(#39=IfcProductDefinitionShape($,$,(#38)))  On instance:     #40=IfcBeam('2mdSQ8KRvFMQtZmTrQAqrb',$,'Example Beam RoundedRectangle',$,'Beam',$,#39,$,$)

</div></div></ul><!--EndFragment-->
</body>
</html>

```

All these `  (((True and False) or (True and 1 == 0) or not True)) ` add more to the confusion than help solving the issue.

**Reply by Moult** [reply]
You can (should? our docs are outdated) replace all ifcopenshell.api.run('foo.bar', ...) with ifcopenshell.api.foo.bar(...). That'll get you autocompletion in your IDE with function signatures. Every API function comes with details docs and examples: https://docs.ifcopenshell.org/autoapi/ifcopenshell/api/index.html

For example https://docs.ifcopenshell.org/autoapi/ifcopenshell/api/context/index.html

Reading the geometry creation section is critical for understanding the concepts of how geometry works in IFC: https://docs.ifcopenshell.org/ifcopenshell-python/geometry_creation.html

I'll update the docs to use the static function calls. That should help :)

**Reply by aothms** [reply]
When you take a look at the VS errors, the most insightful thing will probably be the header

![afbeelding](https://github.com/IfcOpenShell/IfcOpenShell/assets/1096535/69d17684-9001-4eb1-9393-538ff072880e)

Which you can look up in the schema spec

![afbeelding](https://github.com/IfcOpenShell/IfcOpenShell/assets/1096535/14079841-a023-4335-982c-a3945943f0ad)

Docs for IFC4 are here https://standards.buildingsmart.org/IFC/RELEASE/IFC4/ADD2_TC1/HTML/

**Reply by Moult** [reply]
To understand the validation errors, you'll probably need to understand a bit about how IFC is defined. IFC contains "Where Rules" in EXPRESS that define validation requirements. So that validation rule is `IfcProduct.PlacementForShapeRepresentation` which is defined here: https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcProduct.htm#5.1.3.10.4-Formal-propositions

The text description should help make sense of it.

**Reply by zooid** [reply]
Is it not possible for the ifcopenshell to take into consideration these "Where Rules", i.e. the function or the class that creates IfcBeam for example. To ask the user for all the information to properly place the beam and also ask user for profile, brep, etc. I am a Naval Architect by education, so you can imagine how IFC is far from my field.
There could be a function somewhere like ifcopenshell.file() where you start it all, this function could be creating all the necessary things like Project->Site->Building->Storey->GeometricRepresentationContext->etc. everything you need so that applications reading this IFC require as mandatory objects. Then the person writing the app could just focus on the actual geometry.

Don't get me wrong, I appreciate the work done behind ifcopenshell, I can imagine how much work is behind it.
I'm just showing one way to improve it, make it more "approachable" by non-seasoned IFC experts.

**Reply by Moult** [reply]
There isn't anything that is necessary - everything is optional. For example, the site, building, and storey is completely optional. The geometric contexts are optional (FM models may contain no geometry, same with sequencing models or cost schedules).

The API takes into consideration all the where rules - if you use the API you shouldn't be running into validation issues (if you do, file a bug). You're using the low-level manual functions which are probably only recommended for more advanced / manual development.

That said, we do have a few starter templates but they haven't been updated in a while. I'll update them and let you know.

**Reply by zooid** [reply]
I wouldn't say everything is optional since I said, the IFC I was creating was failing the validation because of the topology of the file. Revit could not display the beams because of toplogy. Also, to import IFC in 3DExperience you do need a building and a storey, I'm not sure about the site. I haven't used Tekla so, I can't say anything but I've heard from people using it it also has requirements. It cannot import IFCs generated by 3DExperience IFC exporter.

**Reply by zooid** [reply]
on your question what is not working from the examples:
```
builder = ifcopenshell.util.shape_builder.ShapeBuilder(model)
outer_curve = builder.polyline([(0.,0.), (100.,0.), (100.,50.), (51.2,98.7), (18.5,105.3), (0.,77.5)],
    arc_points=[4], closed=True)
inner_curve = builder.circle((50.,50.), radius=10.)
profile = builder.profile(outer_curve, inner_curves=[inner_curve], name="Arbitrary")
```
this didn't work for me. Arbitrary profile section


Update:
on this page `https://docs.ifcopenshell.org/ifcopenshell-python/geometry_creation.html` all the examples use the `run` function without explanation where it is coming from and this is the first page I opened from the documentation. ;)

**Reply by Moult** [reply]
FYI here is the smallest valid IFC:

```python
import ifcopenshell.api.root
import ifcopenshell.api.project
model = ifcopenshell.api.project.create_file()
ifcopenshell.api.root.create_entity(model, ifc_class="IfcProject", name="My Project")
model.write("/home/dion/minimal.ifc")
```

I'll update the docs to use the static functions which will help all the confusion around the run function.

> this didn't work for me. Arbitrary profile section

I just tested and it works for me. What issue did you have? What error did you get?

**Reply by zooid** [reply]
this is directly from PyCharm without even running it:
![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/1837230/bea38607-a9cd-4d8a-8d5f-15e8790b1d17)

**Reply by Moult** [reply]
Do `import ifcopenshell.util.shape_builder`.

**Reply by zooid** [reply]
Thanks, I'll try, but I'll have to change a few functions to be able to test it.

**Reply by Moult** [reply]
I've rewritten your script to use the API. It should now be valid and hopefully simpler :) https://gist.github.com/Moult/d4a462bed4522bbc200c96a8e80bb6e4 hope it helps!

**Reply by zooid** [reply]
Thanks, alas I got even more functions PyCharm doesn't recognize:
![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/1837230/02bf971a-4cee-4fff-85d9-2ca88f3ead2d)
```
Traceback (most recent call last):
  File "Z:\projects\PY\CPython\ifc_creation\DEBUG_simplified.py", line 355, in <module>
    create_ifc_file("C:\\Temp", "x3d_ifc")
  File "Z:\projects\PY\CPython\ifc_creation\DEBUG_simplified.py", line 276, in create_ifc_file
    model = ifcopenshell.api.project.create_file()
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: module 'ifcopenshell.api.project' has no attribute 'create_file'
```

before you ask:
```
Microsoft Windows [Version 10.0.19045.4529]
(c) Microsoft Corporation. All rights reserved.

(venv) Z:\projects\PY\CPython\ifc_creation>python --version
Python 3.11.9

```
![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/1837230/990c1ce9-6db4-4861-bd06-f25f711f593d)

**Reply by Moult** [reply]
How did you install IfcOpenShell-Python? Perhaps you have an old version? Can you try updating to the latest? BTW feel free to jump on https://osarch.org/chat for a live chat which might be faster, and we can also screenshare if you prefer.

**Reply by Moult** [reply]
OK, 240423 is very old, you can see it says the latest version is 0.7.0240521.

If you want to stick to the old version you can just replace `ifcopenshell.api.foo.bar(...)` with `ifcopenshell.api.run("foo.bar", ...)` - they are equivalent. Or you can just upgrade. The latest version is now 0.7.0.240627.
